streaming/store.go

```diff
@@ -0,0 +1,248 @@
+		JobName:             migrationJobId,
+		JobType:             constants.MINIMAL_DOWNTIME_MIGRATION,
+		Dialect:             conv.SpDialect,
+		Session:             string(sessionBytes),
+		JobMetadata:         string(jobMetadataBytes),
+		SpannerDatabaseName: spannerDatabaseName,
+		CreatedAt:           createTimestamp,
+		UpdatedAt:           createTimestamp,
+	}
+	_, err = client.ReadWriteTransaction(ctx, func(ctx context.Context, txn *spanner.ReadWriteTransaction) error {
+		mutation, err := spanner.InsertStruct("JobDetails", jobDetails)
+		if err != nil {
+			return err
+		}
+		err = txn.BufferWrite([]*spanner.Mutation{mutation})
+		if err != nil {
+			return err
+		}
+		return nil
+	})
+	if err != nil {
+		logger.Log.Error(fmt.Sprintf("can't store generated resources for migration job %s: %v\n", migrationJobId, err))
+		return err
+	}
+	return nil
+}
+
+func writeJobResources(ctx context.Context, migrationJobId string, dataShardId string, dataflowResources internal.DataflowResources, datastreamResources internal.DatastreamResources, gcsResources internal.GcsResources, pubsubResources internal.PubsubResources, monitoringResources internal.MonitoringResources, createTimestamp time.Time, client *spanner.Client) error {
+	datastreamResourcesBytes, err := json.Marshal(datastreamResources)
+	if err != nil {
+		logger.Log.Error(fmt.Sprintf("can't marshal datastream resources for data shard %s: %v\n", dataShardId, err))
+		return err
+	}
+	dataflowResourcesBytes, err := json.Marshal(dataflowResources)
+	if err != nil {
+		logger.Log.Error(fmt.Sprintf("can't marshal dataflow resources for data shard %s: %v\n", dataShardId, err))
+		return err
+	}
+	gcsResourcesBytes, err := json.Marshal(gcsResources)
+	if err != nil {
+		logger.Log.Error(fmt.Sprintf("can't marshal gcs resources for data shard %s: %v\n", dataShardId, err))
+		return err
+	}
+	pubsubResourcesBytes, err := json.Marshal(pubsubResources)
+	if err != nil {
+		logger.Log.Error(fmt.Sprintf("can't marshal pubsub resources for data shard %s: %v\n", dataShardId, err))
+		return err
+	}
+	monitoringResourcesBytes, err := json.Marshal(monitoringResources)
+	if err != nil {
+		logger.Log.Error(fmt.Sprintf("can't marshal monitoring resources for data shard %s: %v\n", dataShardId, err))
+		return err
+	}
+	logger.Log.Debug(fmt.Sprintf("Storing generated resources for data shard %s...\n", dataShardId))
+	_, err = client.ReadWriteTransaction(ctx, func(ctx context.Context, txn *spanner.ReadWriteTransaction) error {
+		datastreamMutation, err := createResourceMutation(migrationJobId, dataShardId, datastreamResources.DatastreamName, constants.DATASTREAM_RESOURCE, datastreamResources.DatastreamName, string(datastreamResourcesBytes))
+		if err != nil {
+			return err
+		}
+		dataflowMutation, err := createResourceMutation(migrationJobId, dataShardId, dataflowResources.JobId, constants.DATAFLOW_RESOURCE, dataflowResources.JobId, string(dataflowResourcesBytes))
+		if err != nil {
+			return err
```

Comment:
We should do a best effort at writing the jobs. Even if one fails, we should log and proceed with the remaining.

Reply:
Why? Currently, all resources are persisted together inside a single transaction (multiple mutations) (refer to Line 217)

Note that we don't fail the migration itself, but we either fail or persist all resources created for a migration as one atomic write.